aslam_cv_frames/src/visual-frame.cc

```diff
@@ -185,6 +243,39 @@ int VisualFrame::getTrackId(size_t index) const {
   CHECK_LT(static_cast<int>(index), track_ids.rows());
   return track_ids.coeff(index, 0);
 }
+const Eigen::Block<Eigen::Matrix4Xd, 4, 1>
+VisualFrame::getSemanticObjectMeasurement(size_t index) const {
+  Eigen::Matrix4Xd& boxes =
+      aslam::channels::get_SEMANTIC_OBJECT_MEASUREMENTS_Data(channels_);
+  CHECK_LT(static_cast<int>(index), boxes.cols());
+  return boxes.block<4, 1>(0, index);
+}
+double VisualFrame::getSemanticObjectMeasurementUncertainty(size_t index) const {
+  Eigen::VectorXd& data =
+      aslam::channels::get_SEMANTIC_OBJECT_MEASUREMENT_UNCERTAINTIES_Data(channels_);
+  CHECK_LT(static_cast<int>(index), data.rows());
+  return data.coeff(index, 0);
+}
+int VisualFrame::getSemanticObjectClassId(size_t index) const {
+  Eigen::VectorXi& class_ids =
+      aslam::channels::get_SEMANTIC_OBJECT_CLASS_IDS_Data(channels_);
+  CHECK_LT(static_cast<int>(index), class_ids.rows());
+  return class_ids.coeff(index, 0);
+}
+const Eigen::MatrixXf VisualFrame::getSemanticObjectDescriptor(size_t index) const {
```

Comment:
so after some reading, i think Eigen::MatrixXf::ColXpr is what we want.
compilation fails if we try to return by reference because descriptors.col(index) is temporary.
I tested with the unit test SetGetSemanticObjectDescriptors in test-visual-frame and ColXpr works. Just that we will need to use .eval() to get the Matrix back.

thoughts?

reference:
https://stackoverflow.com/questions/13782926/eigen-accessing-columns-of-a-matrix-by-reference

Reply:
That's one option and probably the safer one. If you can use the colxpr directly for further lookups and computations (like computing the distance to another descriptor) then this is great, if you need to call eval() every time you actually use the descriptor, it's the same as if you just copy it I guess. Alternatively you could use the ptr option, like for the other descriptor, should work since Eigen is col major by default.

Reply:
So my understanding is that we only call.eval() when we want it to return a matrix type and copying will happen if we assign this to something.
When computing eigen matricies, the intermediate results are converted to temporary expressions, so operations between expressions are still valid, for example I can add ColXpr to another ColXpr.
Looking up a value works with ColXpr, i can do ColXpr(0) and it returns the first element.

The other descriptor returns pointer to a unsigned_char, and uses the Eigen::Map function to return an expression, but it still needs to calculate the rows and cols of this mapping.

So i think it is the same thing, except using.col(i) fills the dimension for you.

Reference:
https://eigen.tuxfamily.org/dox/classEigen_1_1Map.html
https://github.com/ethz-asl/maplab_private/blob/15d7921d0f755be980e42eb0ca604de6a40b71f9/map-structure/localization-summary-map/src/localization-summary-map-creation.cc#L176

Reply:
makes sense, let's go for this solution then.